### ISSUE TITLE
Sik-Ho Tang | Review -- MoCo v3: An Empirical Study of Training Self-Supervised Vision Transformers.

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#140]
Sik-Ho Tang. [Review — MoCo v3: An Empirical Study of Training Self-Supervised Vision Transformers](https://sh-tsang.medium.com/review-moco-v3-an-empirical-study-of-training-self-supervised-vision-transformers-4a1a70c53ecd).

**Comment by NorbertZheng** [comment]
## Overview
Instability Study of [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0) for Self-Supervised Learning.
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/b2172928-c7c7-464d-9f09-1137742ff8d8)
[Vision Transformer (ViT)](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0) Network Architecture (Figure from [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0)).

An Empirical Study of Training Self-Supervised Vision Transformers
MoCo v3, by Facebook AI Research (FAIR)
2021 ICCV, Over 100 Citations
Self-Supervised Learning, Unsupervised Learning, Contrastive Learning, Representation Learning, Image Classification, [Vision Transformer (ViT)](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0)

MoCo v3 is an incremental improvement of [MoCo v1](https://sh-tsang.medium.com/review-moco-momentum-contrast-for-unsupervised-visual-representation-learning-99b590c042a9)/[MoCo v2](https://sh-tsang.medium.com/review-moco-v2-improved-baselines-with-momentum-contrastive-learning-8b03598a0324), studying the instability issue when [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0) is used for self-supervised learning.

**Comment by NorbertZheng** [comment]
## MoCo v3 Using [ResNet](https://towardsdatascience.com/review-resnet-winner-of-ilsvrc-2015-image-classification-localization-detection-e39402bfa5d8) (Before Using [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0))
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/751d6700-81ee-4398-80ab-80938a1238c6)
MoCo v3: PyTorch-like Pseudocode.

 - As the same in [MoCo v2](https://sh-tsang.medium.com/review-moco-v2-improved-baselines-with-momentum-contrastive-learning-8b03598a0324), InfoNCE used in [CPC](https://sh-tsang.medium.com/review-representation-learning-with-contrastive-predictive-coding-cpc-cpcv1-8ec7c6e91a8e), is used as the loss function for training. Larger batch is used to include more negative samples.
 - Different from [MoCo v2](https://sh-tsang.medium.com/review-moco-v2-improved-baselines-with-momentum-contrastive-learning-8b03598a0324), in MoCo v3, the keys that used naturally co-exists in the same batch. The memory queue (memory bank) is abandoned. Thus, this setting is the same as [SimCLR](https://sh-tsang.medium.com/review-simclr-a-simple-framework-for-contrastive-learning-of-visual-representations-5de42ba0bc66).
 - The encoder $f_{q}$ consists of a backbone (e.g., [ResNet](https://towardsdatascience.com/review-resnet-winner-of-ilsvrc-2015-image-classification-localization-detection-e39402bfa5d8), [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0)), a projection head [10], and an extra prediction head [18].
 - The encoder $f_{k}$ has the backbone and projection head, but not the prediction head. $f_{k}$ is updated by the moving-average of $f_{q}$ as in [MoCo](https://sh-tsang.medium.com/review-moco-momentum-contrast-for-unsupervised-visual-representation-learning-99b590c042a9), excluding the prediction head.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/dd4f2dc1-6773-4a64-8dcc-964d25f9d656)
The linear probing accuracy with [ResNet](https://towardsdatascience.com/review-resnet-winner-of-ilsvrc-2015-image-classification-localization-detection-e39402bfa5d8)-50 (R50) on ImageNet.

>By using [ResNet](https://towardsdatascience.com/review-resnet-winner-of-ilsvrc-2015-image-classification-localization-detection-e39402bfa5d8)-50, the improvement here is mainly due to the extra prediction head and large-batch (4096) training.

**Comment by NorbertZheng** [comment]
## Stability Study for Basic Factors When Using [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0)
It is straightforward to replace a [ResNet](https://towardsdatascience.com/review-resnet-winner-of-ilsvrc-2015-image-classification-localization-detection-e39402bfa5d8) backbone with a [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0) backbone. But in practice, a main challenge is the instability of training.

### Batch Size
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/c1685e77-71dc-42e9-97ba-b9468fa35da3)
Training curves of different batch sizes.

A larger batch is also beneficial for accuraacy. A batch of 1k and 2k produces reasonably smooth curves, with 71.5% and 72.6% linear probing accuracy.

>The curve of a 4k batch becomes noticeably unstable: see the "dips". The curve of a 6k batch has worse failure patterns.

### Learning Rate
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/55c5a4bc-208f-49cb-bafc-de0946a62ebc)
Training curves of different learning rates.

When $lr$ is smaller, the training is more stable, but it is prone to under-fitting.

>$lr$=1.5e-4 for this setting has more dips in the curve, and its accuracy is lower. In this regime, the accuracy is determined by stability.

### Optimizer
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/ae791651-123b-4a17-88c3-324820e11ad2)
Training curves of LAMB optimizer.

 - AdamW is the default optimizer. LAMB is studied.
 - Although LAMB can avoid sudden change in the gradients, the negative impact of unreliable gradients is accumulated.

>As a result, authors opt to use AdamW.

**Comment by NorbertZheng** [comment]
## Tricks for Improving Stability
### Random Patch Projection
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/e0fc40ee-56ae-454a-aadc-0e6ce5790730)
Gradient magnitude, shown as relative values for the layer.

It is found that **a sudden change of gradients (a "spike") causes a "dip" in the training curve**.

>The gradient spikes happen earlier in the first layer (patch projection), and are delayed by couples of iterations in the last layers.

Better with residual connection when patch projection???

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/148d6956-1944-4199-8cd7-d2979a15cb2d)
Random vs. learned patch projection.

The instability happens earlier in the shallower layers.
 - The patch projection layer in frozen during training. In other words, a fixed random patch projection layer is used to embed the patches.
 - Random patch projection stabilizes training, with smoother and better training curve. This stability benefits the final accuracy, boosting the accuracy by 1.7% to 73.4%.

### Random Patch Projection on [SimCLR](https://sh-tsang.medium.com/review-simclr-a-simple-framework-for-contrastive-learning-of-visual-representations-5de42ba0bc66) and [BYOL](https://sh-tsang.medium.com/review-byol-bootstrap-your-own-latent-a-new-approach-to-self-supervised-learning-6f770a624441)
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/008c8aa3-1f60-47e9-b391-d6a4e43da9c0)
Random vs. learned patch projection on [SimCLR](https://sh-tsang.medium.com/review-simclr-a-simple-framework-for-contrastive-learning-of-visual-representations-5de42ba0bc66) and [BYOL](https://sh-tsang.medium.com/review-byol-bootstrap-your-own-latent-a-new-approach-to-self-supervised-learning-6f770a624441).

>Random patch projection improves stability in both [SimCLR](https://sh-tsang.medium.com/review-simclr-a-simple-framework-for-contrastive-learning-of-visual-representations-5de42ba0bc66) and [BYOL](https://sh-tsang.medium.com/review-byol-bootstrap-your-own-latent-a-new-approach-to-self-supervised-learning-6f770a624441), and increases the accuracy by 0.8% and 1.3%.

**Comment by NorbertZheng** [comment]
## Experimental Results
### Models
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/1bb6b848-43c6-4c0a-9a9c-0e4c6bd55066)
Configurations of [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0) models.

### Training Time
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/9a0e93a4-6356-4259-90cf-da15be390f4c)
It takes 2.1 hours of training [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0)-B for 100 epochs. [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0)-H takes 9.8 hours per 100 epochs using 512 TPUs.

### Self-Supervised Learning Frameworks
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/6bfd6aa1-d833-4c5b-b06d-bdfb8f31338c)
[ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0)-S/16 and [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0)-B/16 in different self-supervised learning frameworks (ImageNet, linear probing).

 - MoCo v3 has better accuracy on [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0) than other frameworks.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/ed36614c-33e6-4bdf-90c0-27d9a8402b04)
Different Backbones and different framworks.

 - MoCo v3 and [SimCLR](https://sh-tsang.medium.com/review-simclr-a-simple-framework-for-contrastive-learning-of-visual-representations-5de42ba0bc66) are more favorable for [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0)-B than R50.
